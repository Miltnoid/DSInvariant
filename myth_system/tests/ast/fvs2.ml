type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type exp =
  | Unit
  | Var of nat
  | Lam of nat * exp
  | App of exp * exp

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (hd, tl) -> Cons (hd, append tl l2)
;;

let fvs : exp -> list |>
  { Unit => []
  | Var (0) => [0]
  | Var (1) => [1]
  | Lam (0, Unit) => []
  | Lam (0, Var (1)) => [1]
  | App (Unit, Unit) => []
  | App (Var (0), Unit) => [0]
  | App (Unit, Var (1)) => [1]
  | App (Var (0), Var (1)) => [0; 1]
  } = ?
