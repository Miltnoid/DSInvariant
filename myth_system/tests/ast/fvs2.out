(* Generated Wed Dec 31 11:17:18 2014 *)
let fvs : exp -> list =
  let rec f1 (e1:exp) : list =
    match e1 with
      | Unit -> Nil
      | Var n1 -> Cons (n1, Nil)
      | Lam (n1, e2) -> f1 e2
      | App (e2, e3) -> append (f1 e2) (f1 e3)
  in
    f1
;;
